Added code to handle delete session procedure

rsarwad opened the pull request:
"[lte][agw]: 
Added code to handle delete session procedure and added routing logic at mme to s11 message to either sgw_s8 task or to spgw_task, based on federated mode of configuration
## Summary
1. Added code at sgw_s8 task to handle Delete Session Request message
2. Built the gRPC message
3. Added code at sgw_s8 task to handle Delete Session response and delete the contexts and OVS tunnels
4. Added routing logic at mme based federated mode of configuration
5. Moved the routing logic from S6Client.cpp to common file, so that it can be accessed by both mme_app and s6a tasks.

## Test Plan
1. Executed s1ap sanity test suite
2. Executed basic attach and detach scenario over S8 task in Tera VM setup

## Additional Information
This PR include changes of handling of both modify bearer request and delete session procedure.
To view the handling of delete session procedure and routing logic at mme, is available in PR, 
https://github.com/rsarwad/magma/pull/9

This PR is dependent on PR, https://github.com/magma/magma/pull/6046 to insert ovs rules.

pshelar committed:
AGW: MME: GTP-APP: Add support to add S8 tunnels

Add support to match on ingress and egress inbound packets.
New API is added to added to gtp task to add these flows:
openflow_add_s8_tunnel()
openflow_del_s8_tunnel()

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

rsarwad committed:
Added code to send delete session request

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Add code to handle create session response at sgw_s8 task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_csr_rsp_in_s8

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_csr_rsp_in_s8

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Incorporated review comments

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_csr_rsp_in_s8

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Added code to handle modify bearer request in s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Added code to handle modify bearer request in s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_dsr_in_s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge branch 'rsarwad_handle_mbr_in_s8' of https://github.com/rsarwad/magma into rsarwad_handle_dsr_in_s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Added code to handle delete session procedure and added routing logic to send s11 messages to either spgw task or sgw_s8 task based on federated mode of configuration

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Introduced a variable within pdn_context to route s11 messages appropraite task based on federated mode of configuration

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_dsr_in_s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

ulaskozat commented on lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:
```diff
@@ -84,6 +84,10 @@
 extern task_zmq_ctx_t mme_app_task_zmq_ctx;
 extern int pdn_connectivity_delete(emm_context_t* emm_context, pdn_cid_t pid);
 
+static void send_s11_modify_bearer_request(
```
if this is intended for S8 task only then I suggest to use something like `send_modify_bearer_request_to_S8task`.

ulaskozat replied:
never mind I have missed that this is splitting the request among S8 and and SPGW tasks.

rsarwad replied:
No problem

rsarwad committed:
Incorporated review comments

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Adding common utility file to have common functions defined

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_dsr_in_s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_dsr_in_s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Corrected code to send proper IMIESV, ECGI

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

pshelar committed:
AGW: MME: GTP: add S8 TEID args

Add missing TEIDs to program flow from eNB to PGW and vice-versa.
now the API takes in in and out TEID for S1 (eNB) tunnel as well as
S8 (PGW) tunnel.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

ulaskozat commented on lte/gateway/c/oai/tasks/mme_app/mme_app_itti_messaging.c:
```diff
@@ -373,20 +375,23 @@ int mme_app_send_s11_create_session_req(
+  int mode =
+      match_fed_mode_map((char*) session_request_p->imsi.digit, LOG_MME_APP);
+  if (mode == S8_SUBSCRIBER) {
+    OAILOG_INFO_UE(
```
shall we make this debug level?

ulaskozat commented on lte/gateway/c/oai/tasks/mme_app/mme_app_itti_messaging.c:
```diff
@@ -373,20 +375,23 @@ int mme_app_send_s11_create_session_req(
+    send_msg_to_task(&mme_app_task_zmq_ctx, TASK_SGW_S8, message_p);
+    ue_mm_context->pdn_contexts[pdn_cid]->route_s11_messages_to_s8_task = true;
+  } else {
+    OAILOG_INFO_UE(
```
debug level?

ulaskozat commented on lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:
```diff
@@ -4086,3 +4083,30 @@ void mme_app_handle_modify_bearer_rsp(
+      ue_context_p->emm_context._imsi.length);
+
+  if (pdn_context_p->route_s11_messages_to_s8_task) {
+    OAILOG_INFO_UE(
```
debug level?

rsarwad replied:
My understanding on info level would capture the call flow; So I had logged with info, to know where s11 messages would be routed through.
Let me know your thoughts on same

ulaskozat commented on lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:
```diff
@@ -4086,3 +4083,30 @@ void mme_app_handle_modify_bearer_rsp(
+        ue_context_p->mme_ue_s1ap_id);
+    send_msg_to_task(&mme_app_task_zmq_ctx, TASK_SGW_S8, message_p);
+  } else {
+    OAILOG_INFO_UE(
```
debug level?

rsarwad replied:
same as above

ulaskozat commented on lte/gateway/c/oai/lib/s8_proxy/s8_client_api.cpp:
```diff
@@ -101,6 +103,56 @@ static void get_paa_from_proto_msg(
+        context_teid);
+    s8_delete_session_rsp->cause = REMOTE_PEER_NOT_RESPONDING;
+  }
+  OAILOG_INFO_UE(
```
DEBUG level?

ulaskozat commented on lte/gateway/c/oai/lib/s8_proxy/s8_client_api.cpp:
```diff
@@ -371,6 +440,36 @@ static void convert_proto_msg_to_itti_csr(
+    imsi64_t imsi64, Imsi_t imsi, teid_t sgw_s11_teid, teid_t pgw_s5_teid,
+    ebi_t bearer_id) {
+  OAILOG_FUNC_IN(LOG_SGW_S8);
+  OAILOG_INFO_UE(
```
DEBUG level?

ulaskozat commented on lte/gateway/c/oai/tasks/mme_app/mme_app_detach.c:
```diff
@@ -115,10 +116,25 @@ void mme_app_send_delete_session_request(
+      ue_context_p->emm_context._imsi.length);
+
+  if (ue_context_p->pdn_contexts[cid]->route_s11_messages_to_s8_task) {
+    OAILOG_INFO_UE(
```
debug level?

ulaskozat commented on lte/gateway/c/oai/tasks/mme_app/mme_app_detach.c:
```diff
@@ -115,10 +116,25 @@ void mme_app_send_delete_session_request(
+        ue_context_p->mme_teid_s11);
+    send_msg_to_task(&mme_app_task_zmq_ctx, TASK_SGW_S8, message_p);
+  } else {
+    OAILOG_INFO_UE(
```
debug level?

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_dsr_in_s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Modified code to update qos from create session response only valid_qos flag is set

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

pshelar committed:
fix

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

rsarwad committed:
Modified code to use updated ovs api to add tunnel

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_handle_dsr_in_s8_task

Signed-off-by: rashmi <rashmi.sarwad@radisys.com>

ulaskozat reviewed (approved):
LGTM.

ulaskozat commented:
This PR is passing integ tests in dev environment and CI failure looks like a server connection issue. Going to merge it.
